### DISCUSSION TITLE
Arriving at an elegant and typesafe pattern for multi-action routes

### DISCUSSION THREAD

**Discussion by gabimor** [remix-run/remix#5338] [General]
I've been using remix daily for the past 1.5 years now, greatly enjoying the ideas, DX and productivity all around.
Developing in Remix for me is a quest for elegant patterns to ensure quality software, typescript and zod are a big part of the thing. 

One thing i've yet to find a really elegant solution for is multi action routes, specifically:
1. having a multiple actions on a route
2. having good typing when using `useActionData` or `fetcher.data`
3. having good typing in `fetcher.submission` (for optimistic ui)
4. having as little boilerplate as possible with regards to reading the FormData, validating it, etc.
5. having one source of truth for actions types and have it infered (from the action function) rather than written seperately

I've tried a few patterned to get the desired result, but havn't found a nice one that feels covers all bases
@sergiodxa  great `remix-utils` has `named-actions` that really comes close, but the subaction handlers are defined inside the main remix function so can't be used to typing in the components meaning in `useActionData<typeof mySubActionFunc>`, `mySubActionFunc` isn't defined.
i'm also using @kiliman's `remix-params-helper`and `remix-typedjson` and @airjp73 `remix-validated-form` quite a lot

I really feel like this type of scenario is at the core of any non trivial routes and Remix and the community can come up with a nice solid pattern.
My latest iteration on this is something along these lines (note: this almost works with remix-utils, just need to pass ActionArgs to the sub action handlers, a little PR and we're there):

```typescript
export async function action({ request}: ActionArgs) {
  return namedAction(request, {
    saveDocument,
    deleteDocument,
    shareDocument,
    submitDocumentForReview
  })
}

async saveDocument({ request }: ActionArgs){
 // do your specific handler stuff here
}

async deleteDocument({ request }: ActionArgs){
  const parsedFormData = await getFormData(request, myZodDeleteActionSchema)
  // do something here
}

async shareDocument{ request }: ActionArgs){
  // using a remix-validated-form validation 
  const parsedFormData = await myShareDocumentValidator.validate(await request.formData())

  if (!parsedFormData.data) return validationError(parsedFormData.error) 
  
  // do something with parsedFormData.data

  // using remix-typedjson for improved typings
  return typedjson({ success: true })
}

async submitDocumentForReview({ request }: ActionArgs) {
}

function MyComponent() {

  const subActionFetcher = useFetcher<typeof mySubActionHandlerFunction>()
  subActionFetcher.data //  this will be typed with mySubActionHandlerFunction return value
  getFormData<typeof mySubActionHandlerFunction>(subActionFetcher.submission.formData) // this as well

  subActionFetcher.submit(myFormData, {method: "post", action:"/?subActionHandler"})
}
```

Hope this was clear enough,
I'd really love to hear your take on this, what are you using for routes with lot's of sub actions ?

**Comment by kiliman** [comment]
Nice pattern. Thanks for using my packages!

Getting types when you're going to a specific named action is fine, like your `subActionFetcher`. However, going to the main route and using `useActionData` will probably give you a union type. I haven't used the `namedAction` helper, so not positive.

My pattern is typically to redirect on success and only return on validation failures or throw Response on anything else. So I guess if you're using named actions and only return validation errors, then you won't need to worry too much about typing `useActionData` as it should always be the same type: _FormError_ or whatever

**Comment by sergiodxa** [comment]
The `namedAction` function from Remix Utils should give you a TypedResponse with an union of all responses from each sub action.

I recommend you to add a property to the response so you can treat it as a discriminated union and do something like this:

```ts
let actionData = useActionData<typeof action>()

if (actionData.type === "SAVE_DOCUMENT") {
  // do something with a SAVE_DOCUMENT action response
}

// more conditions here
```

**Reply by theocerutti** [reply]
is this the only way to do this?

Because it's seems difficult (or UGLY) to apply condition everywhere.
Example:

```
  useEffect(() => {
    if (
      actionData?.type === "myaction" &&
      actionData?.success
    ) {
      // ....
    }
  }, [actionData?.success]);
```

**Reply by sergiodxa** [reply]
You could abstract that in an assertion function
```ts
function assertIsMyAction(actionData: unknown): asserts actionData is MyAction {
  if ("type" in actionData && actionData.type === "myAction") return
  throw new Error("Action data is not MyAction")
}
```
Then you can do
```ts
assertIsMyAction(actionData)
// here actionData will be typed as MyAction
```

And you can do `Extract<SerializeFrom<typeof action>, { type: "myAction" }>` to get the type of `MyAction`.

**Reply by theocerutti** [reply]
Sounds good! Thanks!

**Reply by redacor** [reply]
@sergiodxa What do you think of a more generic approach?
```typescript
type Action<T> = T extends { intent: infer U } ? U : never;

function assertActionResult<T extends { intent: string }, U extends Action<T>>(
  result: T,
  intent: U,
): asserts result is Extract<T, { intent: U }> {
  if (result.intent !== intent) {
    throw new Error(`Expected intent ${intent}, but got ${result.intent}`);
  }
}
```

In an action:

```TypeScript
return json(
  { success: false, intent, errors } as const,
  { status: 400 }
);

// Later in the same action
return json(
  { success: true, intent, data } as const
);
```

In a component:
```TypeScript
assertActionResult(someActionResult, "some-intent");

// Narrow the type of the action result even further
if (!result.success) {
  // do something with errors
} else {
  // do something with data
}
```

**Comment by gabimor** [comment]
thank you guys for you comments and great packages!
I'll try your suggestions and will update if anything interesting comes up :)

**Comment by subhamX** [comment]
This thing need more recognition.. Thank you for starting such a beautiful thread. 

Q: It has been more than 2 years now. What's your take on building a "type-safe" remix app, with multiple actions handling multiple forms etc.

**Reply by gustavopch** [reply]
This will probably be solved once they implement React Server Functions (https://react.dev/reference/rsc/server-functions).

**Reply by subhamX** [reply]
Thanks for the reply. Do we know when they're planning to do release it? RSC has been around for sometime now. And is now stable I believe.

TBH, these are the particular pain pointers in remix for me.
1. Single action with if-else is a pain. 
2. Also, I hate that loader gets data only from URL params. Having big pagination cursor on address bar looks awful. 

 RSC, in remix would be lit..

**Reply by gustavopch** [reply]
They don't give us ETAs, but my guess is Q1 2025. Agree with the pain points. Regarding the cursor in the URL, you'll still need it if you want people to be able to copy/paste the URL and see the same thing. "Looks awful" isn't a good reason to remove this ability. I don't think the size of the URL even matters. It should only matter for the few URLs that will be used as entry points in marketing.

**Comment by Cerberus** [comment]
I've tried to create a prototype with [trpc](https://trpc.io) (and some patches).

Here is the final result..

![preview](https://github.com/user-attachments/assets/1e7613da-720e-41b6-8f60-6ca20da73a0d)

![remix](https://github.com/user-attachments/assets/e13c50e3-d2c6-4c01-8118-42b31b8a56e9)

Advantages of this implementation:
1. Separate of actions – self-designed
2. Type safe – trpc
3. Single hook for loader & actions – self-designed
4. 2-way params validations (server side & client side) – self-designed

Hope it inspires you.

PS: It's Remix on iOS app with SPA Mode and headless server like normal website. Nothing special.
